### PR TITLE
Feat/cron recommendation status

### PR DESCRIPTION
**Pull request by DorisWongNHS**

### CODE REVIEW

File: lombok.config [hunk 1]
```diff
@@ -0,0 +1,2 @@
+config.stopBubbling = true
+lombok.addLombokGeneratedAnnotation = true
```
Comment (DorisWongNHS) [inline lombok.config:2]: Config file is added to stop sonar complaining the `@data` test coverage 
https://medium.com/@mladen.bolic/lombok-data-improve-your-code-coverage-a74fb624a72b

**Review by JosephKelly — CHANGES_REQUESTED** [review]
Thanks for adding the comment! about Lombok & Sonar! I had wondered.

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceImpl.java [hunk 2]
```diff
@@ -294,6 +296,30 @@ public RecommendationStatus getRecommendationStatusForTrainee(String gmcId) {
...
   }
 
+  /**
+   * Get recommendation status check dtos
+   *
+   * @return A list of submitted to gmc recommendation
+   */
+  public List<RecommendationStatusCheckDto> getRecommendationStatusCheckDtos() {
+    List<RecommendationStatusCheckDto> recommendationStatusCheckDtos = new ArrayList<>();
+    List<Recommendation> recommendations = recommendationRepository
```
[the remark this answers is not in the document]
Comment (DorisWongNHS) [reply application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceImpl.java:306]: Thanks, it is updated to use `getRecommendationStatusCheckDtos().forEach...` 
For the "doctorsForDb.isEmpty()" issue, there are lots of the recommendation would not have a matched doctorsForDB (many of them are migrated from the existing reval). I choose to do nothing for trainees not in doctorsForDB, as the system only count the trainee which is active in GMC which will have a record in doctorsForDB.

File: application/src/main/resources/application.yml [hunk 3]
```diff
@@ -66,14 +66,17 @@ app:
     reval.exchange: ${REVAL_RABBIT_EXCHANGE:reval.exchange}
     reval.queue.recommendation.syncstart: ${REVAL_RABBIT_RECOMMENDATION_SYNC_START_QUEUE:reval.queue.recommendation.syncstart}
     reval.routingKey.recommendation.syncstart: ${REVAL_RABBIT_RECOMMENDATION_SYNCSTART_ROUTING_KEY:reval.recommendation.syncstart}
-
+    reval.exchange.recommendationstatuscheck: ${REVAL_RABBIT_EXCHANGE_RECOMMENDATION_STATUS_CHECK:reval.exchange.recommendationstatuscheck}
```
[the remark this answers is not in the document]
Comment (DorisWongNHS) [reply application/src/main/resources/application.yml:69]: I have removed the line for `reval.exchange.recommendationstatuscheck` as we will use the `reval.exchange` instead.

File: application/src/main/resources/application.yml [hunk 4]
```diff
@@ -66,14 +66,17 @@ app:
...
+    reval.queue.recommendationstatuscheck.requested.gmcclient: ${REVAL_RABBIT_RECOMMENDATION_STATUS_CHECK_REQUESTED_GMCCLIENT_QUEUE:reval.queue.recommendationstatuscheck.requested.gmcclient}
+    reval.routingKey.recommendationstatuscheck.requested: ${REVAL_RABBIT_RECOMMENDATION_STATUS_CHECK_REQUESTED_ROUTING_KEY:reval.recommendationstatuscheck.requested}
 
   gmc:
     url: ${GMC_CONNECT_URL:http://localhost:8090/GMCConnectMock2020/GMCWebServices}
     soapActionBase: ${GMC_SOAP_ACTION_BASE:https://webcache.gmc-uk.org/GMCWebServices/}
     gmcUsername: ${GMC_USER_NAME:guest}
     gmcPassword: ${GMC_PASSWORD:guest}
     designatedBodies: ${DESIGNATED_BODY_CODE:1-AIIDHJ,1-AIIDMQ,1-AIIDNQ,1-AIIDMY,1-AIIDQQ,1-AIIDWT,1-AIIDR8,1-AIIDSA,1-AIIDH1,1-AIIDWA,1-AIIDVS,1-AIIDWI,1-AIIDSI}
+    cronExpression: ${CRON_EXPRESSION:0 0 * * * *}
```
[the remark this answers is not in the document]
Comment (DorisWongNHS) [reply application/src/main/resources/application.yml:79]: I have changed it to use `app.gmc.recommendationstatuscheck.cronExpression`. It contains the keyword of the event, hope it is clearer.

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/GmcClientService.java [hunk 5]
```diff
@@ -67,6 +68,47 @@
   @Value("${app.gmc.soapActionBase}")
   private String gmcSoapBaseAction;
 
+  @Value("${app.rabbit.reval.exchange.recommendationstatuscheck}")
+  private String revalExchangeRecommendationStatus;
+
```
[the remark this answers is not in the document]
Comment (DorisWongNHS) [reply application/src/main/java/uk/nhs/hee/tis/revalidation/service/GmcClientService.java:73]: It is changed to use the `app.rabbit.reval.exchange instead.

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/GmcClientService.java [hunk 6]
```diff
@@ -67,6 +68,47 @@
...
+    this.rabbitTemplate = rabbitTemplate;
+  }
+
+  /**
+   * Cron job to send RecommendationStatusDtos to Rabbit queue for GMC recommendation status check.
+   */
+  @Scheduled(cron = "${app.gmc.cronExpression}")
+  public void sendRecommendationStatusRequestToRabbit() {
+    log.info("Start cron job: sendRecommendationStatusRequestToRabbit()");
+    List<RecommendationStatusCheckDto> recommendations = recommendationService.getRecommendationStatusCheckDtos();
```
[the remark this answers is not in the document]
Comment (DorisWongNHS) [reply application/src/main/java/uk/nhs/hee/tis/revalidation/service/GmcClientService.java:102]: It is updated. Cheers.

**Review by JosephKelly — APPROVED** [review]

**Pull request merged**